sys-boot/edk2/edk2-2018.recipe

```diff
@@ -0,0 +1,89 @@
+SUMMARY="EFI firmware development environment"
+DESCRIPTION="The EDK II project is the response to the EFI community’s \
+request for a better build and version tracking environment for UEFI and PI \
+development. The main difference between the EDK II to the original EDK is \
+the Enhanced Build Environment of the EDK II. The advantages of the Enhanced \
+Build Environment include:
+
+* Operating System independence
+* Flexibility in choosing the compiler and assembler tools
+* The ability to generate working code using open source build tools and \
+applications
+* Enhanced development and build capability of modules and module packages
+* Use of build configuration tools and data sets to provide flexible process
+* Online source control allows users to contribute code and become participants"
+HOMEPAGE="http://www.tianocore.org/"
+COPYRIGHT="2004-2016 Intel Corporation
+	2008-2010 Apple Inc
+	2011-2015 ARM Limited
+	2014-2015 Linaro Limited
+	2013-2015 Red Hat, Inc."
+LICENSE="MIT"
+REVISION="1"
+SOURCE_URI="https://github.com/tianocore/edk2/archive/vUDK$portVersion.tar.gz"
+CHECKSUM_SHA256="a37aa6dc7a4326b562ce03e225afd96a0cd1dc847d9aeddfd867ce6b2e7cc7bc"
+SOURCE_DIR="edk2-vUDK$portVersion"
+PATCHES="edk2-$portVersion.patchset"
+
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
+SECONDARY_ARCHITECTURES="x86"
+DISABLE_SOURCE_PACKAGE="yes"
+
+commandSuffix=$secondaryArchSuffix
+if [ "$targetArchitecture" = x86_gcc2 ]; then
+	commandSuffix=
+fi
+
+PROVIDES="
+	edk2$secondaryArchSuffix = $portVersion compat >= 0
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+ 	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libuuid$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:find
+	cmd:gcc$secondaryArchSuffix
+	cmd:iasl$commandSuffix
+	cmd:make
+	cmd:nasm
+	cmd:python2
+	cmd:sed
+	"
+PATCH()
+{
+	local ACTIVE_PLATFORM="MdeModulePkg/MdeModulePkg.dsc"
+	local TARGET="RELEASE" # DEBUG, RELEASE, NOOPT
+	local TARGET_ARCH="X64" # IA32, IPF, X64, EBC, ARM or AArch64
+	local TOOL_CHAIN_TAG="GCC5"
+
+	if [ "$effectiveTargetArchitecture" = x86 ]; then
+		local TARGET_ARCH="IA32" # IA32, IPF, X64, EBC, ARM or AArch64
+	fi
+
+	sed -i "s|@ACTIVE_PLATFORM@|$ACTIVE_PLATFORM|g" BaseTools/Conf/target.template
+	sed -i "s|@TARGET@|$TARGET|g" BaseTools/Conf/target.template
+	sed -i "s|@TARGET_ARCH@|$TARGET_ARCH|g" BaseTools/Conf/target.template
+	sed -i "s|@TOOL_CHAIN_TAG@|$TOOL_CHAIN_TAG|g" BaseTools/Conf/target.template
+
+	sed -i "s|gcc-ar|ar|g" BaseTools/Conf/tools_def.template
+}
+
+BUILD()
+{
+	make -C BaseTools
+	. ./edksetup.sh
+	local EDK_TOOLS_PATH="$sourceDir"/BaseTools
+	. ./edksetup.sh BaseTools
+	build
+}
+
+INSTALL()
+{
+	mkdir -p "$dataDir"/edk2
+	find Build -type f -name "*.efi" -exec cp "{}" "$dataDir"/edk2 \;
```

Comment:
```diff
-	mkdir -p $dataDir/edk2
+	install -m 0755 -d "$dataDir"/edk2
-	find Build -type f -name "*.efi" -exec cp "{}" $dataDir/edk2 \;
+	cd Build
+	install -m 0644 -t "$dataDir"/edk2 */RELEASE*/*/*.efi
```
There is only one directory which matches this pattern: `$sourceDir/Build/*/RELEASE*/*/`
And that directory has all the `*.efi` we want to install. There are other subdirectories, with a deeper level, but they contain variants (e.g. debug) or copies of the files in the directory mentioned above, but these variants have the same names as the `*.efi` files in that main directory.
So using a `find` to copy everything would lead each target file being overwritten several times. And the last copies could even be from the debug variants.

BTW, the build fails with `gcc 5.4` on x86 (both primary and secondary), even if we define `HOST_ARCH=IA32`. But I should probably upgrade to `gcc 7.3` and try again.

Reply:
@fbrosson If you need to upgrade, then the recipe should declare dependency on newer gcc

Reply:
`gcc 7.3.0` did not help. So I guess we don't need to declare any condition for a particular version of gcc.
But maybe the problem is somewhere else:
```
/boot/system/develop/tools/x86/bin/../lib/gcc/i586-pc-haiku/7.3.0/../../../../i586-pc-haiku/bin/ld: unrecognised emulation mode: elf_i386
Supported emulations: elf_i386_haiku
collect2: error: ld returned 1 exit status
make: *** [/sources/edk2-vUDK2018/Build/MdeModule/RELEASE_GCC5/IA32/MdeModulePkg/Application/HelloWorld/HelloWorld/DEBUG/HelloWorld.dll] Error 1
GNUmakefile:385: recipe for target '/sources/edk2-vUDK2018/Build/MdeModule/RELEASE_GCC5/IA32/MdeModulePkg/Application/HelloWorld/HelloWorld/DEBUG/HelloWorld.dll' failed
```
Why does it try to build a **_DLL_**!!! See the `HelloWorld.dll` at the end of the last line.
I guess some Makefiles need to be changed or skipped.

This is puzzling, however, because on x86_64 I don't get this error. And I do get a `HelloWorld.efi`:
```bash
package list -p edk2-2018-1-x86_64.hpkg | grep -i hello
data/edk2/HelloWorld.efi
```

Reply:
Thanks for the input, but i don't care about 32 bit.

Reply:
I think it might be a good idea to mark x86 as untested instead of broken:
```diff
-ARCHITECTURES="!x86_gcc2 !x86 x86_64"
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
-SECONDARY_ARCHITECTURES="!x86"
+SECONDARY_ARCHITECTURES="?x86"
```
With just this extra change (neutral for x86_64) in `BUILD`:
```diff
-	make -C BaseTools
+	if [ "$effectiveTargetArchitecture" = x86 ]; then
+		HOST_ARCH=IA32
+	fi
+	make -C BaseTools ${HOST_ARCH:+HOST_ARCH=$HOST_ARCH}
```
the build starts for x86 (primary and secondary).
This would allow owners of 32-bit EFI systems to investigate and, maybe, fix the build on x86 :)
What do you think?
(That said, I'm OK with keeping the recipe as is, i.e. with x86 marked as broken.)